lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h

```diff
@@ -340,7 +404,275 @@ class StructFieldAccessInfo {
+    unsigned Id;
+    ExecutionCountType ExecutionCount;
+    DataBytesType DataSize;
+    bool Collapsed;
+    bool LoopArc;
+    Node *FromNode;
+    Node *ToNode;
+    // For Collpased Nodes
+    std::unordered_set<Entry *> CollapsedEntries;
+  };
+
+  /// This structure stores nodes in each basic block and is used to help
+  /// building FRG
+  struct BasicBlockHelperInfo {
+    BasicBlockHelperInfo() : RemainBytes(0), FirstNode(NULL), LastNode(NULL) {}
+    DataBytesType RemainBytes;
+    Node *FirstNode;
+    Node *LastNode;
+    BasicBlockSetType BackEdgeSet;
+  };
+
+public:
+  FieldReferenceGraph(const Function *F) : Func(F), RootNode(NULL) {
+    NodeList.clear();
+    EdgeList.clear();
+    EntryList.clear();
+    BBInfoMap.clear();
+  }
+
+  ~FieldReferenceGraph() {
+    for (auto *N : NodeList)
+      delete N;
+    for (auto *E : EdgeList)
+      delete E;
+    for (auto *E : EntryList)
+      delete E;
+    for (auto &it : BBInfoMap)
+      delete it.second;
+    NodeList.clear();
+    EdgeList.clear();
+    EntryList.clear();
+    BBInfoMap.clear();
+  }
+
+  unsigned getNumNodes() const { return NodeList.size(); }
+
+  Node *getNodeById(unsigned Id) const {
+    assert(Id < NodeList.size());
+    return NodeList[Id];
+  }
+
+  /// Functions for building FRG
+  /// Creator and getter of helper info for the basic block, useful when connect
+  /// nodes from different basic blocks
+  /// %{
+  BasicBlockHelperInfo *createBasicBlockHelperInfo(const BasicBlock *BB);
+  BasicBlockHelperInfo *getBasicBlockHelperInfo(const BasicBlock *BB) const;
+  /// %}
+
+  /// The two functions are used to create a new node in the graph, unconnected
+  /// with other nodes, and return the pointer to the Node
+  /// %{
+  Node *createNewNode(FieldNumType FieldNum, DataBytesType S);
+  Node *createNewNode() { return createNewNode(0, 0); }
+  /// %}
+
+  /// The two functions are used to connect two nodes in FRG with or without
+  /// given weight
+  /// %{
+  void connectNodes(Node *From, Node *To, ExecutionCountType C, DataBytesType D,
+                    bool BackEdge = false);
+  void connectNodes(Node *From, Node *To) { connectNodes(From, To, 0, 0); }
+  /// %}
+
+  /// The getter and setter of entry node in the FRG
+  /// %{
+  Node *getRootNode() const { return RootNode; }
+  void setRootNode(Node *N) { RootNode = N; }
+  /// %}
+
+  /// Functions for collapsing FRG
+  /// Convert a node to a collapsed entry and add it to Edge
+  void collapseNodeToEdge(Node *N, Edge *E);
+
+  /// Copy a collapsed entry to a new edge
+  void moveCollapsedEntryToEdge(Entry *Entry, Edge *FromEdge, Edge *ToEdge);
+
+  /// For debug
+  void debugPrint(raw_ostream &OS) const;
+  void debugPrintCollapsedEntries(raw_ostream &OS,
+                                  FieldReferenceGraph::Edge *E) const;
+
+private:
+  const Function *Func;
+  Node *RootNode;
+  std::vector<Node *> NodeList;
+  std::vector<Edge *> EdgeList;
+  std::vector<Entry *> EntryList;
```

Comment:
Again, it's because I am not using unique_ptr so I use these containers to hold all the pointers I allocated to deallocate them in destructor... Won't need them later if I use unique_ptr. I'll add some comments as suggested.